File: hackerrank/problem solving/algorithms/2 - implementation/45 - modified kaprekar numbers/modified_kaprekar_numbers.cs
// https://www.hackerrank.com/challenges/kaprekar-numbers/problem?isFullScreen=true

int lower = int.Parse(Console.ReadLine()!);
int upper = int.Parse(Console.ReadLine()!);
bool validRange = false;

for (int num = lower; num <= upper; num++)
    if (IsNumberKaprekar(num))
    {
        Console.Write(num + " ");
        validRange = true;
    }

if (!validRange)
    Console.WriteLine("INVALID RANGE");

bool IsNumberKaprekar(int n)
{
    long squareNumber = (long)Math.Pow(n, 2);
    long divisor = (long)Math.Pow(10, NumberDigits(n));
    long leftNumber = Math.DivRem(squareNumber, divisor, out long rightNumber);
    return n == leftNumber + rightNumber;
}

int NumberDigits(long n)
{
    return n.ToString().Length;
}

File: hackerrank/problem solving/algorithms/2 - implementation/45 - modified kaprekar numbers/modified_kaprekar_numbers.ts
// https://www.hackerrank.com/challenges/kaprekar-numbers/problem?isFullScreen=true

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString: string = '';
let inputLines: string[] = [];
let currentLine: number = 0;

process.stdin.on('data', function (inputStdin: string): void {
  inputString += inputStdin;
});

process.stdin.on('end', function (): void {
  inputLines = inputString.split('\n');
  inputString = '';
  main();
});

function readLine(): string {
  return inputLines[currentLine++];
}

//////////////////////////////////////////////////

function main() {
  let validRange = false;

  for (
    let lower = +readLine(), num = lower, upper = +readLine();
    num <= upper;
    num++
  )
    if (isNumberKaprekar(num)) {
      process.stdout.write(num + ' ');
      validRange = true;
    }

  if (!validRange) console.log('INVALID RANGE');
}

function isNumberKaprekar(n: number): boolean {
  const squareNumber = Math.pow(n, 2);
  const divisor = Math.pow(10, numberDigits(n));
  const leftNumber = Math.trunc(squareNumber / divisor);
  const rightNumber = squareNumber % divisor;
  return n === leftNumber + rightNumber;
}

function numberDigits(n: number): number {
  return n.toString().length;
}
